File: tpl/src/rngsse2/examples/examples_SSE_FORTRAN/gm31_sse.f90

PROGRAM GM31_Example

   PARAMETER (N=100000000)
   INTEGER i,GM31_SSE_generate,sum
   REAL val,rsum

   TYPE GM31_state
     INTEGER z(64)
   END TYPE GM31_state
   TYPE(GM31_state) state

   TYPE GM31_sse_state
     INTEGER z(132)
   END TYPE GM31_sse_state
   TYPE(GM31_sse_state) SSE_state

   CALL GM31_init ( state )
   CALL GM31_Get_SSE_state (state,SSE_state)
   CALL GM31_Print_State (state)

   sum = 0

   DO i=1,N
     sum = sum + GM31_SSE_generate (SSE_state)
   END DO

   WRITE(*,1) N
   rsum = sum/4294967296.
   val = GM31_SSE_generate (SSE_state)/4294967296.
   IF (rsum<0) THEN
     rsum=rsum+1
   ENDIF
   IF (val<0) THEN
     val=val+1
   ENDIF
   WRITE(*,2) rsum
   WRITE(*,3) val

1  FORMAT(I9," GM31_SSE pseudorandom numbers generated.")
2  FORMAT("Fractional part of the total sum of generated numbers: ",F8.6)
3  FORMAT("Next output value: ",F8.6)

END PROGRAM GM31_Example

File: tpl/src/rngsse2/examples/examples_SSE_C/gm31_sse.c

#include<stdio.h>
#include<gm31.h>

#define NN       100000000UL

int main(void){ 
   long i; unsigned int sum=0;
   gm31_state state; gm31_sse_state sse_state;
   gm31_init_(&state); gm31_get_sse_state_(&state,&sse_state);
   gm31_print_state_(&state);
   for(i=0;i<NN;i++) sum+=gm31_sse_generate_(&sse_state);
   printf("%ld GM31_SSE pseudorandom numbers generated.\n",NN);
   printf("Fractional part of the total sum of generated numbers: %f\n",sum/4294967296.);
   printf("Next output value: %f\n",gm31_sse_generate_(&sse_state)/4294967296.);
   return 0;
}
